### PR TITLE
feat(hass): use binary sensors when notifications has only two states

### LINKED ISSUE #107: [feat] Support all notification types with custom icon
Related to #105

### LINKED ISSUE #118: [bug] Aeotec Water Sensor 6 does not discover binary sensors necessary for water detection
**Version**

Build/Run method
- [X] Docker
- [ ] PKG
- [ ] Manually built (git clone - npm install - npm run build ) 

zwavejs2mqtt version: 1.2.3

**Describe the bug**
The Aeotec Water Sensor 6 is detected as the correct model and some sensors setup, but the two binary sensors are not discovered by HA for the two water probes (using the dock). This worked correctly in 1.6. 

The two notification classes are detected and sent to mqtt:

When clear (value:0):

```json
{
   "id":"130-113-1-Water Alarm-Sensor status",
   "nodeId":130,
   "commandClass":113,
   "commandClassName":"Notification",
   "endpoint":1,
   "property":"Water Alarm",
   "propertyName":"Water Alarm",
   "propertyKey":"Sensor status",
   "type":"number",
   "readable":true,
   "writeable":false,
   "label":"Sensor status",
   "ccSpecific":{
      "notificationType":5
   },
   "min":0,
   "max":255,
   "list":true,
   "states":[
      {
         "text":"idle",
         "value":0
      },
      {
         "text":"Water leak detected",
         "value":2
      }
   ],
   "value":0,
   "lastUpdate":1609220737810
}
```

When water is detected (value:2):

```json
{
   "id":"130-113-1-Water Alarm-Sensor status",
   "nodeId":130,
   "commandClass":113,
   "commandClassName":"Notification",
   "endpoint":1,
   "property":"Water Alarm",
   "propertyName":"Water Alarm",
   "propertyKey":"Sensor status",
   "type":"number",
   "readable":true,
   "writeable":false,
   "label":"Sensor status",
   "ccSpecific":{
      "notificationType":5
   },
   "min":0,
   "max":255,
   "list":true,
   "states":[
      {
         "text":"idle",
         "value":0
      },
      {
         "text":"Water leak detected",
         "value":2
      }
   ],
   "value":2,
   "lastUpdate":1609220736395
}
```

But binary sensors are not created:

image

Here is the ozw 1.6 device file: https://github.com/OpenZWave/open-zwave/blob/master/config/aeotec/zw122.xml

I have several of these and they are all the same.

**To Reproduce**
1. Pair sensor

**Expected behavior**
Two binary_sensors should be created reflecting the notification command class.

**Additional context**

Here is the log file during a trip of one of the water sensors:

```
05:45:36.112 DRIVER « [Node 130] [REQ] [ApplicationCommand]
                      └─[MultiChannelCCCommandEncapsulation]
                        │ source:      1
                        │ destination: 1
                        └─[NotificationCCReport]
                            notification type:   Water Alarm
                            notification status: 255
                            notification state:  Water leak detected
05:45:36.117 CNTRLR   [Node 130] [+] [Notification] Water Alarm[Sensor status]: 2       [Endpoint 1]
05:45:36.330 SERIAL « 0x0114000400820e600d01017105000000ff0502000082                      (22 bytes)
05:45:36.331 SERIAL » [ACK]                                                                   (0x06)
05:45:36.332 DRIVER « [Node 130] [REQ] [ApplicationCommand]
                      └─[MultiChannelCCCommandEncapsulation]
                        │ source:      1
                        │ destination: 1
                        └─[NotificationCCReport]
                            notification type:   Water Alarm
                            notification status: 255
                            notification state:  Water leak detected
05:45:36.334 CNTRLR   [Node 130] [~] [Notification] Water Alarm[Sensor status]: 2 => 2  [Endpoint 1]
05:45:36.390 SERIAL « 0x0114000400820e600d01007105000000ff0502000083                      (22 bytes)
05:45:36.391 SERIAL » [ACK]                                                                   (0x06)
05:45:36.393 DRIVER « [Node 130] [REQ] [ApplicationCommand]
                      └─[MultiChannelCCCommandEncapsulation]
                        │ source:      1
                        │ destination: 0
                        └─[NotificationCCReport]
                            notification type:   Water Alarm
                            notification status: 255
                            notification state:  Water leak detected
05:45:36.394 CNTRLR   [Node 130] [~] [Notification] Water Alarm[Sensor status]: 2 => 2  [Endpoint 1]
05:45:37.739 SERIAL « 0x0114000400820e600d01017105000000ff0500000080                      (22 bytes)
05:45:37.741 SERIAL » [ACK]                                                                   (0x06)
05:45:37.743 DRIVER « [Node 130] [REQ] [ApplicationCommand]
                      └─[MultiChannelCCCommandEncapsulation]
                        │ source:      1
                        │ destination: 1
                        └─[NotificationCCReport]
                            notification type:   Water Alarm
                            notification status: 255
                            notification event:  Unknown (0x00)
05:45:37.809 CNTRLR   [Node 130] [~] [Notification] Water Alarm[Sensor status]: 2 => 0  [Endpoint 1]
05:45:37.823 SERIAL « 0x0114000400820e600d01007105000000ff0500000081                      (22 bytes)
05:45:37.824 SERIAL » [ACK]                                                                   (0x06)
05:45:37.828 DRIVER « [Node 130] [REQ] [ApplicationCommand]
                      └─[MultiChannelCCCommandEncapsulation]
                        │ source:      1
                        │ destination: 0
                        └─[NotificationCCReport]
                            notification type:   Water Alarm
                            notification status: 255
                            notification event:  Unknown (0x00)

```

### LINKED ISSUE #345: [bug] Fibaro Z-Wave Door / Window Sensor 2 state not exposed to HomeAssistant
Before submitting a bug please read: <https://zwave-js.github.io/zwavejs2mqtt/#/troubleshooting/bug_report>

### Version

Build/Run method

- [x] Docker
- [ ] PKG
- [ ] Manually built (git clone - npm install - npm run build )

zwavejs2mqtt version: 1.0.0-beta.0
zwavejs version: 6.0.0 ?

I use the latest Docker image from last week.

### Describe the bug

Fibaro Z-Wave Door / Window Sensor 2 does not expose state to HomeAssistant.

The state seems to be detected by the ZWaveJS2MQTT but for some reason does not show up in HA.

### To Reproduce

1. Pair a Fibaro Door Sensor 2 to the Z-Wave network
2. Go to Home Assistant > Developer Tools > Entity: sensor.nodeid_13_notification_access_control_door_state
2. Look at data: no status entry present
3. Go to ZWaveJS2MQTT webgui, enable debug logging
4. Open and close door, see state updates in the log just fine

### Expected behavior

Expect the door state to be visible and toggling in HomeAssistant.

### Additional context

Status report within HomeAssistant Developer panel for the entity:
```
id: 13-113-0-Access Control-Door state
nodeId: 13
commandClass: 113
commandClassName: Notification
endpoint: 0
property: Access Control
propertyName: Access Control
propertyKey: Door state
type: number
readable: true
writeable: false
label: Door state
stateless: false
ccSpecific:
  notificationType: 6
min: 0
max: 255
list: true
states:
  - text: Window/door is open
    value: 22
  - text: Window/door is closed
    value: 23
lastUpdate: 1611346323184
nodeName: ''
nodeLocation: ''
friendly_name: Poort State
icon: 'mdi:alarm-light'
```

Debug log within the web-gui of ZWaveJS2MQTT:
```
2021-01-25 19:25:53.055 INFO ZWAVE: Node 13: value updated: 113-0-Access Control-Door state 23 => 22
2021-01-25 19:26:00.626 INFO ZWAVE: Node 13: value updated: 113-0-Access Control-Door state 22 => 23
2021-01-25 19:26:05.461 INFO ZWAVE: Node 13: value updated: 113-0-Access Control-Door state 23 => 22
```

### LINKED ISSUE #374: [feat] HASS: Investigate and support Notification CC as Binary Sensors, when possible
This issue is created to gather all open issues about notifications and try to make them work as binary_sensors in HASS

### PR DESCRIPTION
**Pull request by billiaz**
Try to make Notifications better, byt creating HASS binary sensors.

- [x] cover as much as possible sensors
- [x] Make variables names better
- [x] Possibly improve flow
- [x] Support most sensors
- [x] Feedback

Fixes #374
Fixes #345 
Fixes #118 
Fixes #107 

@blhoward2 Please review this change and Feedback is appreciated
If a device is not properly discovered please paste a Json payload on mqtt

### CODE REVIEW

File: lib/Gateway.js [hunk 1]
```diff
@@ -1610,6 +1637,36 @@ Gateway.prototype.discoverValue = function (node, vId) {
         if (cmdClass === CommandClasses.Basic && valueId.property !== 'event') {
           return
         }
+
+        // Try to define Binary sensor
+        if (valueId.hasOwnProperty('states')) {
```
Comment (robertsLando) [inline lib/Gateway.js:1642]: Use the util method "hasProperty" here

**Review by robertsLando — CHANGES_REQUESTED** [review]
Looks good @billiaz ! :)

File: lib/Gateway.js (hunk 1 above)
Comment (billiaz) [reply lib/Gateway.js]: hasProperty is not working :\

**Comment by blhoward2** [comment]
Ok, I've tested this. I see the new binary sensors under the zwavejs2mqtt front-end for the node (130 in this case), but they never appears in HA. I even tried rediscovering the node and multiple restarts of HA. You'll see more than one startup in the js log because I realized I had mqtt logging turned off. You'll also see that I hate to wake the water sensor a few times but ultimately it went to interview complete.

[zwavejs_1.log](https://github.com/zwave-js/zwavejs2mqtt/files/5900903/zwavejs_1.log)
[zwavejs2mqtt.log.zip](https://github.com/zwave-js/zwavejs2mqtt/files/5900907/zwavejs2mqtt.log.zip)
[node_130.json.zip](https://github.com/zwave-js/zwavejs2mqtt/files/5900909/node_130.json.zip)

**Review by billiaz — COMMENTED** [review]
> Ok, I've tested this. I see the new binary sensors under the zwavejs2mqtt front-end for the node (130 in this case, an Aeotec Water Sensor 6), but they never appears in HA. I even tried rediscovering the node and multiple restarts of HA. You'll see more than one startup in the js log because I realized I had mqtt logging turned off. You'll also see that I had to wake the water sensor a few times but ultimately it went to interview complete.
> 
> [zwavejs_1.log](https://github.com/zwave-js/zwavejs2mqtt/files/5900903/zwavejs_1.log)
> [zwavejs2mqtt.log.zip](https://github.com/zwave-js/zwavejs2mqtt/files/5900907/zwavejs2mqtt.log.zip)
> [node_130.json.zip](https://github.com/zwave-js/zwavejs2mqtt/files/5900909/node_130.json.zip)

I do see the binary sensors are generated for HASS.
I suspect something breaks in hass. to get an entity error frm HASS is good to have DEBUG on (been there!)

```json
        "binary_sensor_cover_status": {
            "type": "binary_sensor",
            "object_id": "cover_status",
            "discovery_payload": {
                "payload_on": 3,
                "payload_off": 0,
                "value_template": "{{ value_json.value }}",
                "device_class": "opening",
                "state_topic": "homeassistant/130/113/0/Home_Security/Cover_status",
                "json_attributes_topic": "homeassistant/130/113/0/Home_Security/Cover_status",
                "device": {
                    "identifiers": [
                        "zwavejs2mqtt_0xeb87ad4a_node130"
                    ],
                    "manufacturer": "AEON Labs",
                    "model": "Water Sensor 6 (ZW122)",
                    "name": "nodeID_130",
                    "sw_version": "1.6"
                },
                "name": "nodeID_130_cover_status",
                "unique_id": "zwavejs2mqtt_0xeb87ad4a_130-113-0-Home_Security-Cover_status"
            },
            "discoveryTopic": "binary_sensor/nodeID_130/cover_status/config",
            "values": [
                "113-0-Home Security-Cover status"
            ],
            "persistent": false,
            "ignoreDiscovery": false
        },
```

another:
```json
        "binary_sensor_water_alarm_1": {
            "type": "binary_sensor",
            "object_id": "water_alarm_1",
            "discovery_payload": {
                "payload_on": 2,
                "payload_off": 0,
                "value_template": "{{ value_json.value }}",
                "device_class": "None",
                "state_topic": "homeassistant/130/113/1/Water_Alarm/Sensor_status",
                "json_attributes_topic": "homeassistant/130/113/1/Water_Alarm/Sensor_status",
                "device": {
                    "identifiers": [
                        "zwavejs2mqtt_0xeb87ad4a_node130"
                    ],
                    "manufacturer": "AEON Labs",
                    "model": "Water Sensor 6 (ZW122)",
                    "name": "nodeID_130",
                    "sw_version": "1.6"
                },
                "name": "nodeID_130_water_alarm_1",
                "unique_id": "zwavejs2mqtt_0xeb87ad4a_130-113-1-Water_Alarm-Sensor_status"
            },
            "discoveryTopic": "binary_sensor/nodeID_130/water_alarm_1/config",
            "values": [
                "113-1-Water Alarm-Sensor status"
            ],
            "persistent": false,
            "ignoreDiscovery": false
        },
```

```json
        "binary_sensor_water_alarm_2": {
            "type": "binary_sensor",
            "object_id": "water_alarm_2",
            "discovery_payload": {
                "payload_on": 2,
                "payload_off": 0,
                "value_template": "{{ value_json.value }}",
                "device_class": "None",
                "state_topic": "homeassistant/130/113/2/Water_Alarm/Sensor_status",
                "json_attributes_topic": "homeassistant/130/113/2/Water_Alarm/Sensor_status",
                "device": {
                    "identifiers": [
                        "zwavejs2mqtt_0xeb87ad4a_node130"
                    ],
                    "manufacturer": "AEON Labs",
                    "model": "Water Sensor 6 (ZW122)",
                    "name": "nodeID_130",
                    "sw_version": "1.6"
                },
                "name": "nodeID_130_water_alarm_2",
                "unique_id": "zwavejs2mqtt_0xeb87ad4a_130-113-2-Water_Alarm-Sensor_status"
            },
            "discoveryTopic": "binary_sensor/nodeID_130/water_alarm_2/config",
            "values": [
                "113-2-Water Alarm-Sensor status"
            ],
            "persistent": false,
            "ignoreDiscovery": false
        }
    },
```

i notice they have class as None! I will tryt o fit a class to make them look nicer.

from this link https://www.home-assistant.io/integrations/binary_sensor/#device-class

I suggest to use "moisture" or "problem" class.
I tend to believe moisture might fit better.

File: lib/Gateway.js [hunk 2]
```diff
@@ -643,6 +643,33 @@ function getMappedStateTemplate (state, defaultValueKey) {
...
+  for (const listKey in states) {
+    map.set(states[listKey].value, states[listKey].text)
+  }
+
+  return map
+}
+function getBinaryPayloadFromSensor (cfg, devicePayload, off = 0) {
+  const stateKeys = Array.from(
+    getMapFromArrayOfStates(devicePayload.states).keys()
+  )
```
Comment (robertsLando) [inline lib/Gateway.js:662]:
```suggestion
  const stateKeys = valueId.states.map(s => s.value)
```

File: lib/Gateway.js [hunk 3]
```diff
@@ -1610,28 +1637,81 @@ Gateway.prototype.discoverValue = function (node, vId) {
...
+                }
+                cfg.object_id = valueId.propertyName
+                break
+              case 'Water Alarm':
+                cfg.discovery_payload.device_class = 'moisture'
+                break
+              default:
+                cfg.discovery_payload.device_class = 'None'
+            }
+            cfg = getBinaryPayloadFromSensor(cfg, valueId, off)
```
Comment (robertsLando) [inline lib/Gateway.js:1692]:
```suggestion
            cfg = setBinaryPayloadFromSensor(cfg, valueId, off)
```

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: lib/Gateway.js (hunk 3 above)
Comment (billiaz) [reply lib/Gateway.js:1700]: Agree, done this change

File: lib/Gateway.js [hunk 4]
```diff
@@ -643,6 +643,33 @@ function getMappedStateTemplate (state, defaultValueKey) {
...
 }
 
+/**
+ * Converts an States Array of maps to map
+ *
+ * @param {array} states States Array of Maps
+ */
+function getMapFromArrayOfStates (states) {
+  const map = new Map()
+  for (const listKey in states) {
```
[the remark this answers is not in the document]
Comment (billiaz) [reply lib/Gateway.js:653]: should be array. of is different? will check it

File: lib/Gateway.js [hunk 5]
```diff
@@ -643,6 +643,33 @@ function getMappedStateTemplate (state, defaultValueKey) {
...
+    map.set(states[listKey].value, states[listKey].text)
+  }
+
+  return map
+}
+function getBinaryPayloadFromSensor (cfg, devicePayload, off = 0) {
```
[the remark this answers is not in the document]
Comment (billiaz) [reply lib/Gateway.js]: made some more changes on this bit

File: lib/Gateway.js (hunk 2 above)
Comment (billiaz) [reply lib/Gateway.js]: done! thanks!

File: lib/Gateway.js [hunk 6]
```diff
@@ -643,6 +643,39 @@ function getMappedStateTemplate (state, defaultValueKey) {
   return `{{ {${map.join(',')}}[value_json.value] | default(${defaultValue}) }}`
 }
 
+/**
+ * Converts an States Array of maps to map
+ *
+ * @param {array} states States Array of Maps
+ */
+function getMapFromArrayOfStates (states) {
```
Comment (robertsLando) [inline lib/Gateway.js:651]: @billiaz This can be removed now

File: lib/Gateway.js [hunk 7]
```diff
@@ -643,6 +643,39 @@ function getMappedStateTemplate (state, defaultValueKey) {
...
+
+  return map
+}
+
+/**
+ * Generates payload for Binary use from a state object
+ *
+ * @param {object} cfg Hass discovery Configuration
+ * @param {object} valueId The device ValueID
+ * @param {number} offStateValue The value number to consider off state
```
Comment (robertsLando) [inline lib/Gateway.js:665]:
```suggestion
 * @param {import('../types').HassDevice} cfg Hass discovery Configuration
 * @param {import('../types').Z2MValueId} valueId The device ValueID
 * @param {number} offStateValue The value number to consider off state
```

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: lib/Gateway.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (billiaz) [reply lib/Gateway.js]: it is an object/map and we extract the index of this. therefore (Reading a bit) for...in is better ?

**Review by robertsLando — DISMISSED** [review]
LGTM. @billiaz Feedback? Is this working?

**Comment by billiaz** [comment]
I got some nice additions from @blhoward2 as two sensors are discovered but as Generic! and might worth include them!
that all! other than this for me it worked like charm! 

@blhoward2 has a separate issue, which is causing his HASS not to discover

**Comment by blhoward2** [comment]
I'll try again tonight, my time. MQTT discovery has been working fine in other contexts so I'm not sure what the issue would be.

**Comment by robertsLando** [comment]
@billiaz This looks good to me, so if @blhoward2 feedback is ok I will merge this or let you do that :)

**Comment by aretakisv** [comment]
I am adding a definition for notification device_classes to cover waterleak
i will put these under Moisture device_class

**Comment by aretakisv** [comment]
@robertsLando it is ready!

**Review by robertsLando — APPROVED** [review]

**Comment by blhoward2** [comment]
@robertsLando @billiaz Hey guys, I think we need to revert this. So contact sensors aren't the only ones with two values sending basic set events. Double tap also works that way and this breaks it. We need to make sure it is actually a water sensor or whatever before we do this change.

**Pull request merged**